Wrappers/Python/cil/processors/PaganinProcessor.py

```diff
@@ -0,0 +1,537 @@
+        (\frac{\mathcal{F}\left ( M^2I_{norm}(x, y,z = \Delta) \right )}{1 + 
+          \alpha\left ( k_x^2 + k_y^2 \right )}  \right )\right ),
+    
+    where
+
+        - :math:`T`, is the sample thickness,
+        - :math:`\mu = \frac{4\pi\beta}{\lambda}` is the material linear 
+        attenuation coefficient where :math:`\beta` is the complex part of the 
+        material refractive index and :math:`\lambda=\frac{hc}{E}` is the probe 
+        wavelength,
+        - :math:`M` is the magnification at the detector,
+        - :math:`I_{norm}` is the input image which is expected to be the 
+        normalised transmission data, 
+        - :math:`\Delta` is the propagation distance,
+        - :math:`\alpha = \frac{\Delta\delta}{\mu}` is a parameter determining 
+        the strength of the filter to be applied in Fourier space where 
+        :math:`\delta` is the real part of the deviation of the material 
+        refractive index from 1 
+        - :math:`k_x, k_y = \left ( \frac{2\pi p}{N_xW}, \frac{2\pi q}{N_yW} 
+        \right )` where :math:`p` and :math:`q` are co-ordinates in a Fourier 
+        mesh in the range :math:`-N_x/2` to :math:`N_x/2` for an image with 
+        size :math:`N_x, N_y` and pixel size :math:`W`.
+    
+    A generalised form of the Paganin phase retrieval method can be called 
+    using :code:`filter_type='generalised_paganin_method'`, which uses the 
+    form of the algorithm described in [2]
+    
+    .. math:: T(x,y) = -\frac{1}{\mu}\ln\left (\mathcal{F}^{-1}\left (\frac{
+        \mathcal{F}\left ( M^2I_{norm}(x, y,z = \Delta) \right )}{1 - \frac{2
+        \alpha}{W^2}\left ( \cos(Wk_x) + \cos(Wk_y) -2 \right )}  \right )
+        \right )
+    
+    The phase retrieval is valid under the following assumptions
+
+        - used with paraxial propagation-induced phase contrast images which 
+        can be assumed to be single-material locally
+        - using intensity data which has been flat field corrected
+        - and under the assumption that the Fresnel number 
+        :math:`F_N = W^2/(\lambda\Delta) >> 1`
+    
+    To apply a filter to images using the Paganin method, call 
+    :code:`full_retrieval=False`. In this case the pre-scaling and conversion 
+    to absorption is not applied so the requirement to supply flat field 
+    corrected intensity data is relaxed,
+    
+    .. math:: I_{filt} = \mathcal{F}^{-1}\left (\frac{\mathcal{F}\left ( 
+        I(x, y,z = \Delta) \right )}
+        {1 - \alpha\left ( k_x^2 + k_y^2 \right )}  \right )
+
+    References
+    ---------
+    - [1] https://doi.org/10.1046/j.1365-2818.2002.01010.x 
+    - [2] https://henke.lbl.gov/optical_constants/getdb2.html
+    - [3] https://iopscience.iop.org/article/10.1088/2040-8986/abbab9
+    With thanks to Rajmund Mokso (DTU) for help with the initial implementation 
+    of the phase retrieval algorithm
+
+    """
+
+    UNIT_LIST = ['m','cm','mm','um']
+    UNIT_MULTIPLIERS = [1.0, 1e-2, 1e-3, 1e-6]
```

Comment:
We've discussed implementing units in a more central place. Maybe that should happen here - we don't need to add interaction with framework at this stage. But it'll mean we don't have to update the API when we do expand it, and we don't have to rely on string matching now. I think you could have something quite light in a `units.py` like:

```
from enum import Enum, auto

class Unit(Enum):
    METRE = auto()
    CENTIMETRE = auto()
    MILLIMETRE = auto()
    MICROMETRE = auto()

UNIT_MULTIPLIERS = {
    Unit.METRE: 1.0,
    Unit.CENTIMETRE: 1e-2,
    Unit.MILLIMETRE: 1e-3,
    Unit.MICROMETRE: 1e-6,
}
```

then:
`processor = PaganinProcessor(delta=1e-6, return_unit=Unit.Metre)`

`self.unit_multiplier = UNIT_MULTIPLIERS[return_units]`



It wouldn't be much more effort to add the energy versions too:
```
class EnergyUnit(Enum):
    KEV = auto()
    MEV = auto()
    EV = auto()
```   
with maybe the input argument as:
`energy=(30, EnergyUnit.KEV)`

Reply:
Good point, I've moved the definitions to units.py now. I think we've made it so it will support an enum or a string (EnergyUnit.keV or 'keV') now so this shouldn't change the Paganin interface.